### ISSUE TITLE
setup.py cannot find vcversall.bat on MSWin 8.1 if installed in user AppData

### ISSUE THREAD

**Issue by dedbd72a-7b0f-43c7-8f93-3c3b22f4385c** [python/cpython#76113]
BPO | [31932](https://bugs.python.org/issue31932)
--- | :---
Nosy | @pfmoore, @tjguk, @merwok, @zware, @zooba, @dstufft
PRs | <li>python/cpython#4249</li>
Superseder | <li>bpo-23246: distutils fails to locate vcvarsall with Visual C++ Compiler for Python</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-11-04.21:40:03.186>
created_at = <Date 2017-11-03.10:44:13.250>
labels = ['type-bug', 'library', 'OS-windows']
title = 'setup.py cannot find vcversall.bat on MSWin 8.1 if installed in user AppData'
updated_at = <Date 2017-11-04.23:25:53.893>
user = 'https://bugs.python.org/laranzu'
```

bugs.python.org fields:
```python
activity = <Date 2017-11-04.23:25:53.893>
actor = 'laranzu'
assignee = 'none'
closed = True
closed_date = <Date 2017-11-04.21:40:03.186>
closer = 'steve.dower'
components = ['Distutils', 'Windows']
creation = <Date 2017-11-03.10:44:13.250>
creator = 'laranzu'
dependencies = []
files = []
hgrepos = []
issue_num = 31932
keywords = ['patch']
message_count = 3.0
messages = ['305470', '305566', '305577']
nosy_count = 8.0
nosy_names = ['paul.moore', 'tim.golden', 'eric.araujo', 'zach.ware', 'steve.dower', 'dstufft', 'laranzu', 'jyrkih']
pr_nums = ['4249']
priority = 'normal'
resolution = 'duplicate'
stage = 'resolved'
status = 'closed'
superseder = '23246'
type = 'behavior'
url = 'https://bugs.python.org/issue31932'
versions = ['Python 2.7']
```

</p></details>

**Comment by dedbd72a-7b0f-43c7-8f93-3c3b22f4385c** [comment]
This was raised in issue bpo-23246 but apparently not addressed at the time.

The Visual C for Python 2.7 tools on my MS Windows 8.1 system installed themselves under the invisible AppData directory because I did not do so as Administrator. Running setup.py to build a (2.7) native extension then fails because it can't find VCVERSALL.BAT.

Fix is a couple of extra lines in find_vcvarsall in msvc9compiler.py to search under AppData if it can't be found in the system location. I will submit a pull request.

**Comment by zooba** [comment]
Duplicate of bpo-23246. AFAIK, nothing has changed since then.

The PR in its current form is not acceptable - there is a totally reliable registry key for finding the compiler, which is better than guessing a path. See the code in setuptools (and also make the case for why we should help people to build extensions without the fixes provided by setuptools, but do it on the existing issue).

**Comment by dedbd72a-7b0f-43c7-8f93-3c3b22f4385c** [comment]
The registry key is not totally reliable for finding the compiler.

What I did: my system is Windows 8.1 64 bit with Python 2.7 64 bit installed. I'd been testing a setup.py on Linux, time to try on MSWin. So I downloaded VCForPython27.msi as recommended by the Python docs, ran it, tried python setup.py build_ext. Failed because could not find VCVARSALL.BAT. Standard MSWin troubleshooting first step: restart. Still no.

Search turned up a freshly installed dir under AppData. After searching through distutils code, stepping through it, and poking around with RegEdit it turned out that while my system has various registry keys for Visual Studio versions from 7.1 through 12.0, even keys for 9.0, but not the particular one that find_vcvarsall wants.

Why not? How can I know? I added two lines of code to msvc9compiler.py and my extension now compiled.

To me, only the file system can be a totally reliable guide as to whether a particular path exists. I agree that it's not practical to wildly guess, or to search everywhere. But this is for Python 2.7, where the configuration has been frozen for all time. There's only a couple of places these files could be.

The additional code I've proposed will make Python continue to work even if the registry is clobbered. It will also make Python continue to work if the user copies the files from another system instead of running the installer. I think this is worthwhile, so ask that you reopen this issue.
